apps/balancer-tools/src/app/apr/(utils)/tokenApr.ts

```diff
@@ -285,3 +321,159 @@ const getImplementationAddress = async (
     return undefined;
   }
 };
+
+async function calculateTokenApr({
+  tokenYield,
+  poolId,
+  chain,
+  tokenAddress,
+  tokenWeight,
+  timeStart,
+  timeEnd,
+}: {
+  tokenYield: number;
+  poolId: string;
+  chain: string;
+  tokenAddress: Address;
+  tokenWeight: number;
+  timeStart: number;
+  timeEnd: number;
+}) {
+  const rangeData = await poolsWithCache.gql(chain).poolSnapshotInRange({
+    poolId,
+    timestamp: generateDateRange(timeStart, timeEnd),
+  });
+
+  let pool: Pool;
+
+  if (rangeData.poolSnapshots.length > 0) {
+    pool = rangeData.poolSnapshots[0].pool;
+  } else {
+    const currentPoolData = await pools.gql(chain).Pool({ poolId });
+    if (!currentPoolData || !currentPoolData.pool) {
+      throw new Error("Pool data is not found or is invalid.");
+    }
+    pool = currentPoolData.pool;
+  }
+
+  if (!pool || !pool.tokens) {
+    return 0;
+  }
+
+  let apr;
+
+  if (
+    pool.poolType === PoolTypeEnum.PHANTOM_STABLE ||
+    (pool.poolType === PoolTypeEnum.WEIGHTED &&
+      (pool.poolTypeVersion ?? 0) >= 2)
+  ) {
+    const isExemptFromYieldProtocolFee = pool?.tokens.find(
+      ({ address }) => address === tokenAddress,
+    )?.isExemptFromYieldProtocolFee;
+
+    if (isExemptFromYieldProtocolFee) {
+      apr = tokenYield;
+    } else {
+      //source: https://github.com/balancer/balancer-sdk/blob/f4879f06289c6f5f9766ead1835f4f4b096ed7dd/balancer-js/src/modules/pools/apr/apr.ts#L134
+      apr = tokenYield * (1 - pool.protocolYieldFeeCache || 0.5);
+    }
+  } else if (
+    pool.poolType === PoolTypeEnum.META_STABLE ||
+    pool.poolType?.includes("Gyro")
+  ) {
+    apr = tokenYield * (pool.protocolYieldFeeCache || 0.5);
+  }
+
+  apr = (apr || 0) * tokenWeight;
+
+  return apr;
+}
+
+async function getTokenWeight(
+  tokenAddress: string,
+  startAtTimestamp: number,
+  endAtTimestamp: number,
+  poolNetwork: string,
+  poolId: string,
+): Promise<number> {
+  const rangeData = await poolsWithCache.gql(poolNetwork).poolSnapshotInRange({
+    poolId,
+    timestamp: generateDateRange(startAtTimestamp, endAtTimestamp),
+  });
+
+  let poolData: Pool;
+
+  if (rangeData.poolSnapshots.length > 0) {
+    poolData = rangeData.poolSnapshots[0].pool;
+  } else {
+    const currentPoolData = await pools.gql(poolNetwork).Pool({ poolId });
+    if (!currentPoolData || !currentPoolData.pool) {
+      throw new Error("Pool data is not found or is invalid.");
+    }
+    poolData = currentPoolData.pool;
+  }
+
+  const poolTokenData = poolData.tokens?.filter((token) => {
+    return token.address !== poolData.address;
+  });
```

Comment:
I think you mean that this is the pool tokens filtering out its own BPT, right? Think your names could help understand more why this is going on here. Perhaps we should also add a comment explaining why ignore it.
```suggestion
    const relevantTokens = poolData.tokens?.filter(token => token.address !== poolData.address) || [];
```